### PR TITLE
refactor: create tiler factories

### LINKED ISSUE #33: reduce code duplication

### PR DESCRIPTION
**Pull request by vincentsarago**
closes #33 

should also solve #78 #77 

⚠️  This is a major refactor. ⚠️ 

This PR introduces 2 new Base Class `TilerFactory` and `MosaicTilerFactory`. Each one creates a router with expected endpoints (/tiles, /point, /tilejson ....). 

I've tried to add comments inside the class to help with the review.

```python
app = FastAPI()

cog = TilerFactory(router_prefix="cog")
app.include_router(cog.router, prefix="/cog", tags=["Cloud Optimized GeoTIFF"])
```

### ToDo
- [ ] write docs 

cc @geospatial-jeff

### CODE REVIEW

File: titiler/dependencies.py [hunk 1]
```diff
@@ -44,201 +48,201 @@ async def request_hash(request: Request) -> str:
...
-class CommonTileParams:
-    """Common Tile parameters."""
+def TMSParams(
+    TileMatrixSetId: TileMatrixSetNames = Query(
+        TileMatrixSetNames.WebMercatorQuad,  # type: ignore
+        description="TileMatrixSet Name (default: 'WebMercatorQuad')",
+    )
+) -> morecantile.TileMatrixSet:
+    """TileMatrixSet Dependency."""
+    return morecantile.tms.get(TileMatrixSetId.name)
```
Comment (vincentsarago) [inline titiler/dependencies.py:58]: This is a simple Callable dependency that returns a morecantil TMS model

File: titiler/dependencies.py [hunk 2]
```diff
@@ -44,201 +48,201 @@ async def request_hash(request: Request) -> str:
...
-
-class CommonMetadataParams:
+def MosaicTMSParams(
+    TileMatrixSetId: MosaicTileMatrixSetNames = Query(
+        MosaicTileMatrixSetNames.WebMercatorQuad,  # type: ignore
+        description="TileMatrixSet Name (default: 'WebMercatorQuad')",
+    )
+) -> morecantile.TileMatrixSet:
+    """TileMatrixSet Dependency."""
+    return morecantile.tms.get(TileMatrixSetId.name)
```
Comment (vincentsarago) [inline titiler/dependencies.py:68]: Mosaic only allow WebMercator for now!

File: titiler/dependencies.py [hunk 3]
```diff
@@ -44,201 +48,201 @@ async def request_hash(request: Request) -> str:
...
+
+
+@dataclass
+class DefaultDependency:
+    """Dependency Base Class"""
+
+    kwargs: dict = field(init=False, default_factory=dict)
```
Comment (vincentsarago) [inline titiler/dependencies.py:75]: The DefaultDependency class is more a placeholder than a real dependency for now!

File: titiler/dependencies.py [hunk 4]
```diff
@@ -44,201 +48,201 @@ async def request_hash(request: Request) -> str:
...
+
+
+@dataclass
```
Comment (vincentsarago) [inline titiler/dependencies.py:78]: we now use dataclass for dependency definition

File: titiler/dependencies.py [hunk 5]
```diff
@@ -44,201 +48,201 @@ async def request_hash(request: Request) -> str:
...
+        if parsed.scheme.split("+")[-1] == "mosaicid":
+            self.url = f"{settings.DEFAULT_MOSAIC_BACKEND}{settings.DEFAULT_MOSAIC_HOST}/{parsed.netloc}.json.gz"
+
+        # if parsed.scheme.startswith("landsat+"):
+        #     self.reader = L8Reader
+        #     self.url = self.url.replace("landsat+", "")
+        # elif parsed.scheme.startswith("sentinel2cog+"):
+        #     self.reader = Sentinel2COG
+        #     self.url = self.url.replace("sentinel2cog+", "")
+        # ....
```
Comment (vincentsarago) [inline titiler/dependencies.py:100]: The PathParams is maybe the most important dependency. I've added a `reader` placeholder to add more fancy things like `landsat+mosaicid`

File: titiler/dependencies.py [hunk 6]
```diff
@@ -44,201 +48,201 @@ async def request_hash(request: Request) -> str:
...
+        None,
+        title="Asset indexes",
+        description="comma (',') delimited asset names (might not be an available options of some readers)",
+    )
+    kwargs: dict = field(init=False, default_factory=dict)
+
+    def __post_init__(self):
+        """Post Init."""
+        if self.assets is not None:
+            self.kwargs["assets"] = self.assets.split(",")
```
Comment (vincentsarago) [inline titiler/dependencies.py:117]: ⚠️ we use kwargs to store the `assets` option so we can just pass `**params.kwargs` to the function (in the tiler) and we do not have to worry if the reader accept the asset keyword. 

in the tiler factory with use `add_asset_deps` to specify if we use `AssetsParams` or `DefaultDependency`. The deps will populate the open api with assets (if set to true) but on the function we just need to declare 

```python
        optionsParams = AssetsParams if self.add_asset_deps else DefaultDependency
        
        def info(
            src_path=Depends(self.path_dependency), 
            options: optionsParams = Depends()
        ):
            """Return basic info."""
            with reader(src_path.url) as src_dst:
                info = src_dst.info(**options.kwargs)

```

File: titiler/endpoints/demo.py [hunk 7]
```diff
@@ -1,15 +1,16 @@
...
 
 from starlette.requests import Request
 from starlette.responses import HTMLResponse
 from starlette.templating import Jinja2Templates
 
 router = APIRouter()
-
-
-templates = Jinja2Templates(directory="titiler/templates")
+template_dir = pkg_resources.resource_filename("titiler", "templates")
```
Comment (vincentsarago) [inline titiler/endpoints/demo.py:12]: fixes #77

File: titiler/main.py [hunk 8]
```diff
@@ -18,26 +24,40 @@
...
+
+
+@cog.router.get("/validate", response_model=RioCogeoInfo)
+def cog_validate(
+    src_path: PathParams = Depends(),
+    strict: bool = Query(False, description="Treat warnings as errors"),
+):
+    """Validate a COG"""
+    return rio_cogeo_info(src_path.url, strict=strict)
+
```
Comment (vincentsarago) [inline titiler/main.py:44]: The above is equivalend of extending the TilerFactory Class

```python
from dataclasses import dataclass

@dataclass
class CustomTiler(TilerFactory):
    router_prefix = "cog"
    def __post_init__(self):
        super().__post_init__()
        self._add_validation()
    def _add_validation(self):
        @self.router.get(
            "/validate",
            response_model=RioCogeoInfo,
            name=f"{self.router_prefix}info",
        )
        def validate(
            src_path: PathParams = Depends(),
            strict: bool = Query(False, description="Treat warnings as errors"),
        ):
            """Validate a COG"""
            return rio_cogeo_info(src_path.url, strict=strict)

cog = CustomTiler()
app.include_router(cog.router, prefix="/cog", tags=["Cloud Optimized GeoTIFF"])
```

File: titiler/dependencies.py (hunk 3 above)
Comment (geospatial-jeff) [reply titiler/dependencies.py:75]: If its not meant to be instantiated maybe make it an `ABC`

File: titiler/dependencies.py [hunk 9]
```diff
@@ -44,201 +48,201 @@ async def request_hash(request: Request) -> str:
...
+
+@dataclass
+class DefaultDependency:
+    """Dependency Base Class"""
+
+    kwargs: dict = field(init=False, default_factory=dict)
+
+
+@dataclass
+class PathParams(DefaultDependency):
```
Comment (geospatial-jeff) [inline titiler/dependencies.py:79]: Should this be called `QueryParams` instead?  I don't see any path parameters defined.

File: titiler/dependencies.py [hunk 10]
```diff
@@ -44,201 +48,201 @@ async def request_hash(request: Request) -> str:
...
+    """Create dataset path from args"""
+
+    url: str = Query(..., description="Dataset URL")
+    reader: Optional[Type[BaseReader]] = field(init=False, default=None)  # Placeholder
+
+    def __post_init__(self,):
```
Comment (geospatial-jeff) [inline titiler/dependencies.py:85]: Using `__post_init__` here is serving the same purpose as defining a dependency (ex. [`TMSParams`](https://github.com/developmentseed/titiler/pull/79/files#diff-7e64e756c81e1d4282046f8561c647ebR51-R55)) just with a different pattern.  Would it be possible to use either one or the other everywhere you need to modify a request parameter after it is declared?

Comment (vincentsarago) [reply titiler/dependencies.py:85]: Yeah except if we go ahead with `reader` defined as a scheme (e.g `landsat+...`) It's not really used right now but the Factory expects the PathParams class to have 2 properties (url+reader).

File: titiler/dependencies.py (hunk 9 above)
Comment (vincentsarago) [reply titiler/dependencies.py:79]: @geospatial-jeff I'm not sure to understand

Comment (geospatial-jeff) [reply titiler/dependencies.py:79]: The class is called `PathParams` but it only defines query parameters below (`url: str = Query(...)`).  So this class doesn't seem to be related to path parameters at all.

Comment (geospatial-jeff) [reply titiler/dependencies.py:79]: Oh I see, its called `PathParams` because it creates the dataset path, and I'm confusing that with http path params

**Pull request merged**